tests/neo4j/sessionrun.py

```diff
@@ -207,6 +207,54 @@ def test_iteration_larger_than_fetch_size(self):
             rec = result.next()
             self.assertEqual(rec, exp)
 
+    def test_partial_iteration(self):
+        # Verifies that not consuming all records works
+        self._session1 = self._driver.session("r", fetchSize=2)
+        result = self._session1.run("UNWIND RANGE(0, 1000) AS x RETURN x")
+        for x in range(0, 4):
+            exp = types.Record(values=[types.CypherInt(x)])
+            rec = result.next()
+            self.assertEqual(rec, exp)
+        self._session1.close()
+        self._session1 = None
+
+        # not consumed all records & starting a new session
+        self._session1 = self._driver.session("r", fetchSize=2)
+        result = self._session1.run("UNWIND RANGE(2000, 3000) AS x RETURN x")
+        for x in range(2000, 2004):
+            exp = types.Record(values=[types.CypherInt(x)])
+            rec = result.next()
+            self.assertEqual(rec, exp)
+
+        # not consumed all records & reusing the previous session
+        result = self._session1.run("UNWIND RANGE(4000, 5000) AS x RETURN x")
+        for x in range(4000, 4004):
+            exp = types.Record(values=[types.CypherInt(x)])
+            rec = result.next()
+            self.assertEqual(rec, exp)
+
+        self._session1.close()
+        self._session1 = None
+
+    def test_session_reuse(self):
+        def _test():
+            self._session1 = self._driver.session("r", fetchSize=2)
+            result = self._session1.run("UNWIND [1, 2, 3, 4] AS x RETURN x")
+            if consume:
+                result.consume()
+            result = self._session1.run("UNWIND [5,6,7,8] AS x RETURN x")
+            records = list(map(lambda record: record.values, result))
+            self.assertEqual(records,
+                             [[i] for i in map(types.CypherInt, range(5, 9))])
+            summary = result.consume()
+            self.assertIsInstance(summary, types.Summary)
+            self._session1.close()
+            self._session1 = None
+
+        for consume in (True, False):
+            with self.subTest("consume" if consume else "no consume"):
+                _test()
```

Comment:
The readability of the code could improve with the `consume` flag sent by param to the `_test` function, instead of getting it from the context.

Reply:
Yes and no. I'd rather avoid shadowing outer variables. So I'd have to call it something else inside the function. Which do you prefer?
 1. ```python
    def foo(bar_):
        if bar_:
           pass  # do something
   
    bar = True  # or whatever
    foo(bar)
    ```
 2. ```python
    def foo():
        if bar:
            pass  # do something
    
    bar = True  # or whatever
    foo()
    ```

Reply:
Maybe:

```python
def _test(consume):
    if consume:
        pass # do something

 
with self.subTest("consume"):
    _test(consume=True)

with self.subTest("no consume"):
    _test(consume=False)
```

Or use the _consume in the `for _consume in (True, False):`

But I agree it's a tough decision.

Reply:
I should just transition TestKit to pytest asap. It would make all these problems go away as it has proper test parametrization support. Anyway, until then, I'd leave it as is unless you definitely want me to change it. Regarding your 1st suggestions: looks the cleanest, admittedly.,but it falls short if you want to parametrize more than one thing.